lib/elasticsearch/drain/cli.rb

```diff
@@ -13,24 +13,73 @@ class CLI < ::Thor
       option :asg, required: true
       option :region, required: true
       option :nodes
+      option :number
+      option :continue, type: :boolean, default: true
       def asg # rubocop:disable Metrics/MethodLength
         @drainer = Elasticsearch::Drain.new(options[:host],
                                             options[:asg],
                                             options[:region])
+
         ensure_cluster_healthy
         @active_nodes = drainer.active_nodes_in_asg
 
+        # If :nodes are specified, :number has no effect
+        if options[:nodes]
+          number_to_drain = nil
+          currently_draining_nodes = nil
+        else
+          number_to_drain = options[:number]
+          currently_draining_nodes = drainer.cluster.currently_draining('_id')
+        end
+
+
         # If a node or nodes are specified, only drain the requested node(s)
         @active_nodes = active_nodes.find_all do |n|
           instance_id = drainer.asg.instance(n.ipaddress).instance_id
           options[:nodes].split(',').include?(instance_id)
         end if options[:nodes]
 
         do_exit { say_status 'Complete', 'Nothing to do', :green } if active_nodes.empty?
-        say_status 'Found Nodes', "AutoScalingGroup: #{instances}", :magenta
-        ensure_cluster_healthy
-        drain_nodes
-        remove_nodes
+        say_status 'Found Nodes', "AutoScalingGroup: #{instances(active_nodes)}", :magenta
+
+        until active_nodes.empty?
+          ensure_cluster_healthy
+
+          nodes = active_nodes
+
+          # If there are nodes in cluster settings "transient.cluster.routing.allocation.exclude"
+          # test if those nodes are still in the ASG. If so, work on them first unless nodes are
+          # specified.
+          if currently_draining_nodes
+            nodes_to_drain = active_nodes.find_all { |n| currently_draining_nodes.split(',').include?(n.id) }
+
+            # If the list of nodes_to_drain isn't empty, we want to set nodes to the list of nodes
+            # we've already been working on.
+            unless nodes_to_drain.empty?
+              nodes = nodes_to_drain
+
+              say_status 'Active Nodes', "Resuming drain process on #{instances(nodes)}", :magenta
+            end
+
+            # We should only process currently_draining_nodes once
+            currently_draining_nodes = nil
+          end
+
+          # If we specify a number but DON'T specify nodes, sample the active_nodes.
+          if number_to_drain
+            nodes = nodes.sample(number_to_drain.to_i)
+            say_status 'Active Nodes', "Sampled #{number_to_drain} nodes and got #{instances(nodes)}", :magenta
+          end
+
+          @active_nodes = nodes unless options[:continue]
+
+          drain_nodes(nodes)
+          remove_nodes(nodes)
+
+          # Remove the drained nodes from the list of active_nodes
+          @active_nodes -= nodes
```

Comment:
I started hacking on varying the sleep time so we don't slam the cluster in #26, feel free to take that if it's useful.